[REEF-362] Implement a Wake Transport using HTTP

nhne opened the pull request:
Additional implementation for Wake via HTTP
* new package `org.apache.reef.wake.remote.transport.netty.http`
* followed default implementation of netty
* Added TransportHttpTest
* currently tested only for HTTP

JIRA:
[REEF-362](https://issues.apache.org/jira/browse/REEF-362)

Pull Request
This closes #

DifferentSC commented:
@nhne How about adding tests on `RemoteManager` using `HttpMessagingTransportFactory`?

nhne commented:
Added one Test on `RemoteManager` with `HttpMessagingTransportFactory`.

taegeonum commented:
@nhne I've skimmed this PR, but it looks like it has a bunch of duplicated codes. Can't we minimize the duplicated codes? Maybe we need to refactor the original code to do this.

nhne commented:
I implmeneted in this way to avoid modify the original code. I just worried I can insert err code into original code and compatibility issues. I think refactoring the original code can be eliminate many duplicated code since It does not have much difference. Do you think what would be better?

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/RemoteConfiguration.java:
```diff
@@ -132,4 +139,12 @@ private RemoteConfiguration() {
+  /**
+   * Option for use http.
+   */
+  @NamedParameter(doc = "Option for use http.", default_value = "" + PROTOCOL_NETTY)
```
Converting `String` using blank string and `+` is a bad practice. Use `String.valueOf` instead.

nhne replied:
It was used across the original source. Is it better to update to `String.valueOf` every use?

nhne replied:
It would be better to make another issue for refactoring the whole project.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/TransportFactory.java:
```diff
@@ -82,5 +101,27 @@ Transport newInstance(final String hostAddress,
+   * @return transport
+   */
+  Transport newInstance(final String hostAddress,
+                        int port,
```
why no final here?

nhne replied:
It would be refactored during another PR. Thank you!

taegeonum replied:
You don't have to refactor it in another PR because it is a new code. Please just add `final` here.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/MessagingTransportFactory.java:
```diff
@@ -104,6 +104,33 @@ public Transport newInstance(final String hostAddress,
+                               final int retryTimeout,
+                               final int protocol) {
+    try {
+      TcpPortProvider tcpPortProvider = Tang.Factory.getTang().newInjector().getInstance(TcpPortProvider.class);
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelFutureListener.java:
```diff
@@ -0,0 +1,43 @@
+import io.netty.channel.ChannelFutureListener;
+import org.apache.reef.wake.remote.transport.LinkListener;
+
+class NettyChannelFutureListener<T> implements ChannelFutureListener {
```
This class should be public & final if there is no class extending it.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,68 @@
    */
   public static final int MAXFRAMELENGTH = 10 * 1024 * 1024;
   private final NettyChannelHandlerFactory handlerFactory;
+  private final SslContext sslContext;
```
I think explanations on these variables are necessary.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,68 @@
   public static final int MAXFRAMELENGTH = 10 * 1024 * 1024;
   private final NettyChannelHandlerFactory handlerFactory;
+  private final SslContext sslContext;
+  private final int type;
```
add comments.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+    if (msg instanceof HttpResponse) {
+      System.out.println("CLIENT : IT IS HTTP RESPONSE");
```
Using `LOG` is better practice than using `System.out`. Avoid printing to `System.out` or `System.err` on your PR.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+    if (msg instanceof HttpResponse) {
+      System.out.println("CLIENT : IT IS HTTP RESPONSE");
+    } else if (msg instanceof HttpContent) {
+      System.out.println("CLIENT : It is HttpContent");
```
Use `LOG` instead.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+      System.out.println("CLIENT : IT IS HTTP RESPONSE");
+    } else if (msg instanceof HttpContent) {
+      System.out.println("CLIENT : It is HttpContent");
+      HttpContent httpContent = (HttpContent) msg;
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+    } else if (msg instanceof HttpContent) {
+      System.out.println("CLIENT : It is HttpContent");
+      HttpContent httpContent = (HttpContent) msg;
+      ByteBuf content = httpContent.content();
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+
+      if (msg instanceof LastHttpContent) {
+        final Channel channel = ctx.channel();
+        byte[] message = new byte[content.readableBytes()];
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  @Override
+  protected TransportEvent getTransportEvent(final byte[] message, final Channel channel) {
```
I'm not sure why `protected` is necessary here. It seems that you don't have any extending classes.

nhne replied:
I think this method should not be used outside of netty implementation since the `stage` is declared in `protected`

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  @Override
+  protected void exceptionCleanup(final ChannelHandlerContext ctx, final Throwable cause) {
```
Same here.

nhne replied:
Replied above. I don't know why It says oudated...

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+    final HttpHeaders headers;
+    final HttpRequest request;
```
As only one of these two variables will be used, I think it would be better to define each variable inside each `if` and `else if`.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+
+      if (msg instanceof LastHttpContent) {
+        buf.append("END OF CONTENT\r\n");
+        LastHttpContent trailer = (LastHttpContent) msg;
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+      appendDecoderResult(buf, request);
+    } else if (msg instanceof HttpContent) {
+      LOG.log(Level.FINEST, "HttpContent received");
+      HttpContent httpContent = (HttpContent) msg;
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+    } else if (msg instanceof HttpContent) {
+      LOG.log(Level.FINEST, "HttpContent received");
+      HttpContent httpContent = (HttpContent) msg;
+      ByteBuf content = httpContent.content();
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+      this.httpRequest = request;
+      headers = request.headers();
+      if (!headers.isEmpty()) {
+        for (Map.Entry<String, String> h: headers) {
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+      headers = request.headers();
+      if (!headers.isEmpty()) {
+        for (Map.Entry<String, String> h: headers) {
+          CharSequence key = h.getKey();
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+      if (!headers.isEmpty()) {
+        for (Map.Entry<String, String> h: headers) {
+          CharSequence key = h.getKey();
+          CharSequence value = h.getValue();
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        final Channel channel = ctx.channel();
+        byte[] message = new byte[content.readableBytes()];
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  @Override
+  protected void exceptionCleanup(final ChannelHandlerContext ctx, final Throwable cause) {
```
Why no operation here?

nhne replied:
It is required method to extend `AbstractNettyEventyListener` and it only performs close channel when it is clientListener.

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,30 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
+
+    if (this.uri != null) {
+      try {
+        FullHttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uri.getRawPath());
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,30 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
+        if (listener !=  null) {
+          future.addListener(new NettyChannelFutureListener<>(message, listener));
+        }
+      } catch (InterruptedException ex) {
```
final

DifferentSC commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -95,8 +101,11 @@
+  private final AbstractNettyEventListener clientEventListener;
+  private final AbstractNettyEventListener serverEventListener;
+
+  private final boolean ssl = System.getProperty("ssl") != null;
```
I don't think initializing `final` variable here is good, because it is not a static variable. How about moving this into the constructor?

nhne commented:
@DifferentSC Thank you for reviewing my work! I'll check the things you reviewed.

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,97 @@
+    } else if (msg instanceof HttpContent) {
+      final HttpContent httpContent = (HttpContent) msg;
+      final ByteBuf content = httpContent.content();
+      if (content.isReadable()) {
```
Also, do we ever have to reset the content of the `buf`?

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+    } else {
+      LOG.log(Level.SEVERE, "Unknown type of message received: {0}", msg);
+    }
+    LOG.log(Level.FINEST, buf.toString());
```
don't call `.toString()` prematurely. Better do
```java
LOG.log(Level.FINEST, "Message received: {0}", buf);
```

motus replied:
Also, same as above: if we use `buf` for logging purposes only, make sure you append to it only when the logging level is low enough.

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,155 @@
+
+    buf.append(".. WITH DECODER FAILURE: ");
+    buf.append(result.cause());
+    buf.append("\r\n");
```
why not just
```java
if (!result.isSuccess()) {
  buf.append(".. WITH DECODER FAILURE: ").append(result.cause()).append("\r\n");
}
```

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,30 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
+        request.headers().set(HttpHeaders.Names.ACCEPT_ENCODING, HttpHeaders.Values.GZIP);
+        request.headers().set(HttpHeaders.Names.CONTENT_TYPE, "application/wake-transport");
+        final ByteBuf buf = Unpooled.copiedBuffer(encoder.encode(message));
+        request.headers().set(HttpHeaders.Names.CONTENT_LENGTH, buf.readableBytes());
```
`.set()` calls can be chained:
```java
request.headers()
  .set(HttpHeaders.Names.HOST, uri.getHost())
  .set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.KEEP_ALIVE)
  .set(HttpHeaders.Names.ACCEPT_ENCODING, HttpHeaders.Values.GZIP)
  .set(HttpHeaders.Names.CONTENT_TYPE, "application/wake-transport")
  .set(HttpHeaders.Names.CONTENT_LENGTH, buf.readableBytes());
```

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,30 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
+        request.content().clear().writeBytes(buf);
+        final ChannelFuture future = channel.writeAndFlush(request);
+        future.sync();
+        if (listener !=  null) {
```
extra space before `null` 😄

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,30 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
+      }
+    } else {
+      final ChannelFuture future = channel.writeAndFlush(Unpooled.wrappedBuffer(encoder.encode(message)));
+      if (listener !=  null) {
```
😄

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -79,6 +81,10 @@
 
+  private static final int HANDLER_NETTY = 100;
+  private static final int HANDLER_HTTP_SERVER = 101;
+  private static final int HANDLER_HTTP_CLIENT = 102;
```
Don't we already have these constants in `NettyChannelInitializer`? Let's have a shared `enum` instead

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -95,8 +101,11 @@
+  private final AbstractNettyEventListener clientEventListener;
+  private final AbstractNettyEventListener serverEventListener;
+
+  private final boolean ssl;
```
use naming appropriate for a Boolean flag, e.g. `hasSsl` or `sslSupported` or something

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +141,46 @@ private NettyMessagingTransport(
+    final SslContext sslContextClient;
+    final SslContext sslContextServer;
+
+    ssl = System.getProperty("ssl") != null;
```
is it possible to have, say, `"false"` or `"off"` as a property value?

nhne replied:
It would be good to provide an option for use whether use SSL or not. I would redesign the implementation after refactoring the Wake Transport. Thank you!

nhne replied:
Oh, I misunderstood your saying! Currently, I couldn't find any details about System.getProperty("ssl") except Netty Http example I referenced. I'll find out is there any good implementation for support SSL. Thank you very much for pointing out this!

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +141,46 @@ private NettyMessagingTransport(
+
+    ssl = System.getProperty("ssl") != null;
+
+    if(protocolType == PROTOCOL_HTTP) {
```
space after `if`

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +141,46 @@ private NettyMessagingTransport(
+          sslContextClient = SslContext.newServerContext(ssc.certificate(), ssc.privateKey());
+          sslContextServer = SslContext.newClientContext();
+          LOG.log(Level.FINE, "SSL context created");
+        } catch (Exception ex) {
```
`final`

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +141,46 @@ private NettyMessagingTransport(
+        sslContextClient = null;
+        sslContextServer = null;
+      }
+      this.uri = URI.create(ssl ? "https://" : "http://" + hostAddress);
```
**BUG!**
`?:` has lower priority than `+`. Use parenthesis: `URI.create((ssl ? "https://" : "http://") + hostAddress)`

nhne replied:
Oops... Thank you!!

motus commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,162 @@
+        new InetSocketAddress(hostAddress, port),
+        new ObjectSerializableCodec<String>(),
+        new LoggingLinkListener<String>());
+    link.write(new String("hello1"));
```
no need for `new String(...)` - `"hello1"` is a string already

nhne replied:
It would be changed after refactoring code. Thank you.

motus commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,162 @@
+        new ObjectSerializableCodec<String>(),
+        new LoggingLinkListener<String>());
+    link.write(new String("hello1"));
+    link.write(new String("hello2"));
```
same as above

motus commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,162 @@
+
+  @Test
+  public void testHttpTransportTestEvent() throws Exception {
+    System.out.println(LOG_PREFIX + name.getMethodName());
```
use logging instead

motus commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,162 @@
+    link.write(new TestEvent("hello2", 1.0));
+
+    monitor.mwait();
+    transport.close();
```
`Transport` implements `AutoCloseable` - can use it in `try` block instead of calling `.close()` explicitly

motus commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,162 @@
+    private final Codec<T> codec;
+    private final Monitor monitor;
+    private final int expected;
+    private AtomicInteger count = new AtomicInteger(0);
```
final

nhne commented:
@motus I appreciate for your kind and precise review! I'll check out all about these. thanks a lot!

DifferentSC commented:
Thanks a lot for your detailed comments, @motus! We have decided to split this issue into two, by making a new JIRA issue for refactoring and changing styles of `Transport`. @nhne will work on implementing HTTP transport after the refactoring is done. Thanks!

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -108,23 +150,3 @@ public String toString() {
     return "NettyLink: " + channel; // Channel has good .toString() implementation
   }
 }
-
```
It is separted into another file "NettyChannelFutureListener.java"

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/TransportFactory.java:
```diff
@@ -71,6 +71,25 @@ Transport newInstance(final String hostAddress, int port,
+                        final EStage<TransportEvent> serverStage,
+                        final int numberOfTries,
+                        final int retryTimeout,
+                        final String protocol);
```
How about using enum?

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/MessagingTransportFactory.java:
```diff
@@ -104,6 +105,33 @@ public Transport newInstance(final String hostAddress,
+                               final EStage<TransportEvent> serverStage,
+                               final int numberOfTries,
+                               final int retryTimeout,
+                               final String protocol) {
```
this variable is not used

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelFutureListener.java:
```diff
@@ -0,0 +1,46 @@
+public final class NettyChannelFutureListener<T> implements ChannelFutureListener {
+
+  private final T message;
+  private LinkListener<T> listener;
```
final

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,77 @@
+
+  /* Types for initiating channel */
+  public enum ChannelType {
+    NETTY, HTTP_SERVER, HTTP_CLIENT
```
What is NETTY? Is it a tcp or udp socket channel? The naming is unclear to me because HTTP server and client also uses Netty.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,77 @@
+
+  NettyChannelInitializer(
+      final NettyChannelHandlerFactory handlerFactory,
+      final SslContext sslContext,
```
If this can be null, how about creating another constructor not receiving this variable?

nhne replied:
Unfortunately, current implementation must be changed into `if` statement if take advantage of new constructor.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,97 @@
+          LOG.log(Level.FINEST, "MessageEvent: local: {0} remote: {1} :: {2}", new Object[]{
+                  channel.localAddress(), channel.remoteAddress(), buf});
+        }
+
```
Don't we have to create a new StringBuilder here? because it is the end of the message and we need a new StringBuilder for another response?

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,97 @@
+    } else if (msg instanceof HttpContent) {
+      final HttpContent httpContent = (HttpContent) msg;
+      final ByteBuf content = httpContent.content();
+      if (content.isReadable() && LOG.isLoggable(Level.FINEST)) {
```
Change the order to diminish the number of `content.isReadable()` function calls

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.logging.Logger;
+
+/**
+ * A Netty event listener for server.
```
server -> client?

nhne replied:
Oops, my mistake! Thank you!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,153 @@
+      this.httpRequest = request;
+
+      if (!headers.isEmpty()) {
+        for (final Map.Entry<String, String> h: headers) {
```
h (space):

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,153 @@
+          }
+          buf.append("\r\n");
+        }
+
```
Same as the above. Don't we have to create a new StringBuillder here?

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLink.java:
```diff
@@ -77,9 +96,32 @@ public NettyLink(final Channel channel, final Encoder<? super T> encoder, final
-    if (listener !=  null) {
-      future.addListener(new NettyChannelFutureListener<>(message, listener));
+
+    if (this.uri != null) {
```
I think creating a `NettyHttpLink` would be better instead of fixing this class. 
We have to check whenever `.write` is called whether the uri is null or not. This doesn't look good to me.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +138,46 @@ private NettyMessagingTransport(
-    this.clientEventListener = new NettyClientEventListener(this.addrToLinkRefMap, clientStage);
-    this.serverEventListener = new NettyServerEventListener(this.addrToLinkRefMap, serverStage);
+    if (protocolType.equals(PROTOCOL_NETTY)) {
+      this.clientEventListener = new NettyClientEventListener(this.addrToLinkRefMap, clientStage);
```
How about receiving client and server event listener from the constructor? 
We can remove `if` ... `else` if we inject them in the constructor.

taegeonum replied:
I've discussed it with @nhne in offline, and we decided not to refactor the constructor because there is a little benefit of refactoring it.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -111,6 +117,7 @@
+   * @param protocolType  the protocol to use for transport
    */
   @Inject
   private NettyMessagingTransport(
```
This class contains lots of if else. I'm not sure this is a good design. We can remove the `if`...`else` by receiving variables from the constructor.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/RemoteConfiguration.java:
```diff
@@ -47,6 +47,13 @@
+   *  Unique protocol numbers for choosing protocols.
+   */
+  public static final int PROTOCOL_NETTY = 100;
+  public static final int PROTOCOL_HTTP = 101;
```
@nhne You can convert enum to string. Please see https://stackoverflow.com/questions/6667243/using-enum-values-as-string-literals

nhne commented:
@taegeonum Thanks for sharing detailed review! I'll check all you mentioned above. Thank you.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyLinkFactory.java:
```diff
@@ -0,0 +1,49 @@
+   * @param listener the listener
+   */
+  Link<T> newInstance(final Channel channel,
+                           final Encoder<? super T> encoder,
```
Pleas align the lines

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java:
```diff
@@ -0,0 +1,136 @@
+
+  @Override
+  public String toString() {
+    return "NettyLink: " + channel; // Channel has good .toString() implementation
```
Please use a StringBuffer

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java:
```diff
@@ -0,0 +1,136 @@
+ * If you set a {@code LinkListener<T>}, it keeps message until writeAndFlush operation completes
+ * and notifies whether the sent message transferred successfully through the listener.
+ */
+public class NettyHttpLink<T> implements Link<T> {
```
final

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,98 @@
+          buf.setLength(0); // clearing the buffer
+        }
+
+        if (message.length > 0) {
```
Could you explain when does the message length is zero? and don't we send the message if the length is zero?

nhne replied:
By currrent implementation, Server does not send any response to client. Actually `NettyHttpEventListener` does not receive any response now. But if server return the response to client, there can be HttpResponse with empty message.

We can choose to clear the `channelRead` method to do nothing or leave it for latter usage.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/RemoteConfiguration.java:
```diff
@@ -47,6 +48,14 @@
+   */
+  public static final String PROTOCOL_TCP = ProtocolTypes.TCP.name();
+  public static final String PROTOCOL_HTTP = ProtocolTypes.HTTP.name();
+  public static final String PROTOCOL_HTTPS = ProtocolTypes.HTTPS.name();
```
Do we handle supporting https in this PR?

nhne replied:
I Think features about HTTPS should be handled in another PR. Thank you!

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyDefaultLinkFactory.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Factory that creates a NettyLink.
+ */
+public final class NettyDefaultLinkFactory<T> implements NettyLinkFactory {
```
I've created NettyHttpLinkFactory. Thank you!

nhne commented:
@taegeonum I just pushed a commit reflexing your review. Thank you for your kind review!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,77 @@
+
+  NettyChannelInitializer(
+      final NettyChannelHandlerFactory handlerFactory,
+      final SslContext sslContext,
```
This is for HTTPS?

nhne replied:
I'll remove the whole SSL parts. thank you!

taegeonum commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/RemoteManagerTestHttp.java:
```diff
@@ -0,0 +1,490 @@
+/**
+ * Tests for RemoteManagerFactory with HTTP Transport implementation.
+ */
+public class RemoteManagerTestHttp {
```
Could you please refactor the test codes?

taegeonum commented on lang/java/reef-wake/wake/src/test/java/org/apache/reef/wake/test/remote/TransportHttpTest.java:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Tests for Transport via http.
+ */
+public class TransportHttpTest {
```
Could you please refactor the test codes? you can use the same test cases by injecting different netty transports

nhne replied:
I'm very sorry but I cannot fully understand the saying 'injecting different netty transports'. Could you kindly please explain the direction I should refactor in more detail?

taegeonum replied:
There are lots of duplicate codes in `TransportHttpTest` and `TransportTest`, so I thought we need to refactor it. 
But, it is an separate issue, so let's refactor it in another issue :) Could you please create an issue for it ?

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +136,30 @@ private NettyMessagingTransport(
+    // for HTTP and default Netty
+    sslContextClient = null;
+    sslContextServer = null;
+    if (protocolType.equals(PROTOCOL_HTTP)) {
```
do `protocolType == ProtocolTypes.HTTP`. Don't use strings

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +136,30 @@ private NettyMessagingTransport(
     this.retryTimeout = retryTimeout;
-    this.clientEventListener = new NettyClientEventListener(this.addrToLinkRefMap, clientStage);
-    this.serverEventListener = new NettyServerEventListener(this.addrToLinkRefMap, serverStage);
+    if (protocolType.equals(PROTOCOL_TCP)) {
```
same as above

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -146,15 +172,17 @@ private NettyMessagingTransport(
         .handler(new NettyChannelInitializer(new NettyDefaultChannelHandlerFactory("client",
-            this.clientChannelGroup, this.clientEventListener)))
+            this.clientChannelGroup, this.clientEventListener), sslContextClient,
+                protocolType.equals(PROTOCOL_TCP) ? ChannelType.TCP : ChannelType.HTTP_CLIENT))
```
Implement a method in `ChannelType` to get the channel type by protocol, so you can call e.g. `ChannelType.Client(protocolType)`

motus commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -146,15 +172,17 @@ private NettyMessagingTransport(
         .childHandler(new NettyChannelInitializer(new NettyDefaultChannelHandlerFactory("server",
-            this.serverChannelGroup, this.serverEventListener)))
+            this.serverChannelGroup, this.serverEventListener), sslContextServer,
+                protocolType.equals(PROTOCOL_TCP) ? ChannelType.TCP : ChannelType.HTTP_SERVER))
```
Same as above, e.g. `ChannelType.Server(protocolType)`

nhne commented:
@taegeonum Sorry for late response. I was busy for work about GSoC and reserve force traing(예비군 훈련). I'll check your review now. Thank you for review!

Thank you for good suggestion! We have discussed the using `enum` in parameter issue with @DifferentSC and @taegeonum. Unfortunately, current REEF Tang does not support enum correctly by now. So for now, I implemented in the way that interface uses enum but, uses string internally.
I'll check about other reviews now. Thanks a lot @motus !

nhne commented:
This is error when I tried to apply enum as Tang parameter.
```
Caused by: org.apache.reef.tang.exceptions.NameResolutionException: Could not resolve TCP.  Search ended at prefix TCP This can happen due to typos in class names that are passed as strings, or because Tang is configured to use a classloader other than the one that generated the class reference (check your classpath and the code that instantiated Tang)
at org.apache.reef.tang.implementation.java.ClassHierarchyImpl.getAlreadyBoundNode(ClassHierarchyImpl.java:274)
	at org.apache.reef.tang.implementation.java.ClassHierarchyImpl.getNode(ClassHierarchyImpl.java:261)
	at org.apache.reef.tang.implementation.java.ClassHierarchyImpl.parse(ClassHierarchyImpl.java:171)
	... 35 more
```

motus commented:
Oh, that's very unfortunate. I was sure that Tang has support for enums by now! Luckily, this is easy to fix: here's my pull request #1370 - I hope it will help to make your code a few lines shorter 😄

nhne commented:
Thank you @motus ! I'll fix this after pulling the master. Thanks you very much!

bgchun commented:
@nhne Did you test this code on a YARN cluster? If not, please do so.

nhne commented:
I have ran tests on a YARN clutser. It passed 36 tests successfully. I will keep testing on Yarn by commits.

bgchun commented:
I think it's better to refactor the tests in this PR.

taegeonum commented:
@bgchun Either way is fine with me :) so @nhne, could you please refactor the tests?

nhne commented:
I have changed implementation of `TransportFactory` to use Tang for choosing protocols. And I also merged `TransportHttpTest` into `TransportTest`. It was hard to avoid using duplicated codes in separated test because JUnit didn't allowed multiple constructor for any test class and I thought changing `tpFactory` by a set method is not a good idea.

@taegeonum I would be grateful if you share your idea! Thank you.

taegeonum commented:
@nhne I think you could do the same refactoring on `RemoteManagerTest.java`, as you did on `TransportTest` :)

nhne commented:
I did merged Http tests into `TransportTest` and `RemoteManagerTest` repectly.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,98 @@
+      }
+    } else if (msg instanceof HttpContent) {
+      final HttpContent httpContent = (HttpContent) msg;
+      final ByteBuf content = httpContent.content();
```
This variable is only used when it is `LastHttpContent`. If the httpContent is fragmented, this will be discarded. Is this intended?

taegeonum replied:
We can just get `FullHttpResponse` if we use `HttpObjectAggregator`.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,156 @@
+
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+    if(msg instanceof HttpRequest) {
```
We can just get FullHttpRequest if we use HttpObjectAggregator.

nhne commented:
Thanks for your review! @taegeonum. I'll reconsider the structure of Http Channels considering your review.

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,70 @@
+      } else {
+        ch.pipeline()
+            .addLast("codec", new HttpClientCodec())
+            .addLast("decompressor", new HttpContentDecompressor())
```
I think this is not a necessary component. thank you!

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -24,29 +24,70 @@
+    default:
+      throw new IllegalArgumentException("Invalid type of channel");
+    }
   }
```
I think it is good idea! I'll implement it and test it.

nhne commented:
@taegeonum Sorry for very late commit! I have addressed the review. Would you please have a look?

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,92 @@
+
+      content.readBytes(message);
+      if (LOG.isLoggable(Level.FINEST)) {
+        buf.append("CONTENT: ").append(content.toString(CharsetUtil.UTF_8)).append("\r\n");
```
Why don't you define `buf` as a local variable? It doesn't look like this should be a member variable.

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java:
```diff
@@ -0,0 +1,138 @@
+          .set(HttpHeaders.Names.CONTENT_LENGTH, buf.readableBytes());
+      request.content().clear().writeBytes(buf);
+      final ChannelFuture future = channel.writeAndFlush(request);
+      future.sync();
```
Why sync here?

nhne replied:
I wanted to throw exception when the result of future is failure by `sync()`. but i found its order is weird because listener must be added before its sync!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,151 @@
+ */
+final class NettyHttpServerEventListener extends AbstractNettyEventListener {
+
+  private final StringBuilder buf = new StringBuilder();
```
Same as above. Do we have to define it as a member variable?

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyChannelInitializer.java:
```diff
@@ -19,34 +19,77 @@
+      if(isServer) {
+        pipeline
+            .addLast("codec", new HttpServerCodec())
+            .addLast("requestDecoder", new HttpRequestDecoder())
```
I followed example in the Netty Http server. But by reference document, I think we do need additional codec since `HttpServerCodec` is combination of them. I'll remove and test about it.

nhne commented:
My former implementation need buf as globla variable, but now we don't have to keep it global, so I edited this.
Also I fixed misusing of codec in `NettyChannelInitializer` and sync in `NettyHttpLink`.
@taegeonum Thank you for your review!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpClientEventListener.java:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
```
How about the following code? 
```java
public void channelRead(final ChannelHandlerContext ctx, final FullHttpResponse response) {

	byte[] content = null;
	ByteBuf byteBuf = response.content();
	if (byteBuf.hasArray()) {
		content = byteBuf.array();
	} else {
		content = new byte[byteBuf.readableBytes()];
		byteBuf.readBytes(content);
	}
        .... 
}
```

nhne replied:
I'm sorry that as method channelRead is overriding method, I cannot change type of parameter.
But i'll adapt the name change and other things! Thank you

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
```
Same as the `ClientEventListener` codes

nhne commented:
Sorry for late check. I have committed about your review. Thank you very much!

nhne commented:
I thought about this somewhile... If any listener expect any response without any parameter, intetionally empty request can be sent. Then, this check would be erroneous. So I removed the if statements.

Thank you! @taegeonum

nhne commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java:
```diff
@@ -0,0 +1,138 @@
+      if (listener != null) {
+        future.addListener(new NettyChannelFutureListener<>(message, listener));
+      }
+      future.sync();
```
It is because sync() will throw an Exception when sending request is Interrupted. To catch a failure on sending request, sync was used here.

taegeonum replied:
I mean, why do we need sync? you already registered the future listener and it doesn't look we need sync()

nhne replied:
`listener` can be null in constructor. If no listener's are attached, there would be no warning about failure on sending request. In my opinion, if default listener for failure detect is provided, we don't have to use `sync()`.

taegeonum replied:
Using `Future` and `listener` is to asynchronously write the requests, but if you do `sync()` immediately, I think there is no benefit of the asynchronous write.

nhne replied:
I stronly agree with you. Then what about change the implmentation to adopt default listener for detect failure on complete? or just delete `sync()`? Can I ask which would be better choice?

taegeonum replied:
Sorry, I don't fully understand your question. You've already registered the future listener in line 104.

nhne replied:
@taegeonum the `listener` can be null. The constructor of `NettyHttpLink` at line59 makes listener null. So There can be no listener for catching failure if `sync()` is not performed.

taegeonum replied:
@nhne I see. I think it is okay to me. If we want to ignore the message futures, we can skip the listener. Otherwise, we provide the listener and handle the message futures. So, how about just removing `sync()`?

nhne replied:
Then I'll remove `sync()`. User may provide listener if they need any failure detection. Thank you for sharing your opinion!

nhne commented:
@taegeonum I have changed copiedBuffer into wrappedBuffer. Also I commented about sync upper.
Thank you for your review!

nhne commented:
@taegeonum I have removed `sync()`. Thank you for sharing your opinion!

nhne commented:
@taegeonum Thank you for sharing your idea for `sync()`. I have removed it from `NettyHttpLink`. Thank you!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java:
```diff
@@ -0,0 +1,133 @@
+   */
+  @Override
+  public void write(final T message) {
+    LOG.log(Level.FINEST, "write {0} :: {1}", new Object[] {channel, message});
```
Could you please add the following code? 
```
if (LOG.isLoggable(Level.FINEST)) {
```

nhne commented:
@taegeonum fixed the `LOG` check. Thank you!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -130,10 +132,25 @@ private NettyMessagingTransport(
+
+    // for HTTP and default Netty
+    if (protocolType == ProtocolType.HTTP) {
+      this.uri = URI.create("http://" + hostAddress);
```
@nhne `hostAddress` -> `host`
I've got an exception message when I tested it on my local machine, even though the test passed  
```
Caused by: java.lang.IllegalArgumentException: Illegal character in fragment at index 8: http://##UNKNOWN##
	at java.net.URI.create(URI.java:852)
	at org.apache.reef.wake.remote.transport.netty.NettyMessagingTransport.<init>(NettyMessagingTransport.java:139)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.apache.reef.tang.implementation.java.InjectorImpl.injectFromPlan(InjectorImpl.java:637)
	... 41 more
```

nhne replied:
Sorry for late check. I had to reconfigure my environment to reproduce the error. I'll fix it and push a commit now!

nhne commented:
@taegeonum I have fixed `hostAddress` into `host`. also I found `URI.create` can cause an `IllegalArgumentException` on invalid address, so I added RemoteRuntimeException about this invalid URI.
Thank you for finding this error!

taegeonum commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -139,7 +139,7 @@ private NettyMessagingTransport(
       try{
         this.uri = URI.create("http://" + host);
       } catch (IllegalArgumentException e){
```
Could you please add `final` ?

taegeonum commented:
@nhne LGTM except for the last comment :)

nhne commented:
@taegeonum I have pushed a commit reflecting your review. Thank you!

bgchun commented:
@nhne @taegeonum Thanks for working on this issue. Could you also make sure that this PR works correctly in distributed environments?
